Fixes for libs on MacOS

mtikekar opened the pull request:
- $ORIGIN changed to @loader_path
- libpython's extension is .dylib, no .so
- add install_names to vpi and vhpi extensions
- add libbpython's path to rpath

<!--

Thanks for improving cocotb! Here are some points to make this as smooth as possible.
Not all of them may be applicable.

Most important: please explain *why* you are proposing this change.

* Make sure you have read https://github.com/cocotb/cocotb/blob/master/CONTRIBUTING.md
* Extend or add a test under `tests/test_cases/`.
* Add documentation under `documentation/source/`,
  docstrings in Python code, or Doxygen markup in C/C++ code.
  Use ``versionadded``/``versionchanged``/``deprecated``.
* Add a newsfragment - see `documentation/source/newsfragments/README.rst`.
* Use `closes #XXXX` to auto-close the issue that this PR fixes (if such).

-->

themperek commented:
What error/issue does this solve?

mtikekar commented:
I could not use cocotb at all without all these fixes. The first error is when running `make SIM=icarus` or `make SIM=verilator` -

```
dyld: Library not loaded: @rpath/libpython3.7m.dylib
  Referenced from: .../lib/python3.7/site-packages/cocotb/libs/libcocotb.so
  Reason: image not found
```

This is fixed by adding `sys.prefix + "/lib"` to rpath.

The next error was in cocotb/share/lib/embed/gpi_embed.cpp: "Failed to find python shared library" which was due to `-DPYTHON_SO_LIB=libpython3.7m.so` instead of `libpython3.7m.dylib`.

The third error was that the simulator was able to link to libcocotbvpi_verilator.so but could not run because the install_name was not set on the .so file. I'm not sure if other simulators need it, but it shouldn't hurt.

The change from `$ORIGIN` to `@loader_path` is just cleanup. `$ORIGIN` is not understood on MacOS.

I don't think these changes should affect Windows or Linux, but I will be checking on Linux soon.

cmarqu commented on cocotb/_build_libs.py:
```diff
@@ -174,6 +170,8 @@ def _get_python_lib():
     if os.name == "nt":
         python_lib = _get_python_lib_link() + "." + _get_lib_ext_name()
+    elif sys.platform == "darwin":
+        python_lib = "lib" + _get_python_lib_link() + ".dylib"
```
You could just extend ``_get_lib_ext_name()`` instead, right?

mtikekar replied:
Yes, I could do that. But I found a few places where the default extension for libraries was assumed to be .so (Linux, Mac) or .dll (Windows). I wasn't sure I caught them all.

themperek commented:
@mtikekar So there are few issues here. This would be sperate PRs or minimum sperate commits with message that include information what is the issue this commit solves.

themperek commented:
Aby idea why Travis does not run for this PR?

mtikekar commented:
I can do separate commits on the same PR.

mtikekar commented:
Travis says "abuse detected" for this and my earlier PR. Some discussion [here](https://travis-ci.community/t/some-pr-are-not-analyzed-with-message-abuse-detected/1191/7) and [here](https://travis-ci.community/t/abuse-detected-for-my-prs-and-some-commits/6124/3).

themperek commented on cocotb/_build_libs.py:
```diff
@@ -46,13 +46,7 @@ def build_extension(self, ext):
-                "install_name_tool",
-                "-add_rpath",
-                "@loader_path/libs",
+                sys.prefix + "/lib",
```
Should not be better?
```suggestion
                sysconfig.get_config_var("LIBDIR"),
```

themperek replied:
Possibly we can add this only on `libcocotb` and remove this here?

mtikekar replied:
Yes, I am familiar with `sysconfig`.

themperek commented:
Interesting is the fact that it seems to be possible to do `-Wl,-install_name,@rpath/abc.so` and  `-Wl,-rpath,@loader_path` same time. I have tried different combinations to make it work but not this one (I do not use and do not have Mac).

mtikekar commented:
On Mac, the `install_name` of a library is used to resolve the library at load time. For example, if libA has install_name of `/path/to/libA.dylib` and libB is linked with `-lA`, libB will list `/path/to/libA.dylib` as its dependency (unlike Linux which will list `libA`).

The concept of `rpath` is the same on Linux and Mac, except that instead of `$ORIGIN`, you have to use `@load_path` or `@executable_path`as values for `rpath`.

If we set `-Wl,-rpath` correctly, we can remove the `install_name_tool` invocation.

themperek commented on cocotb/_build_libs.py:
```diff
@@ -148,6 +142,8 @@ def _extra_link_args(lib_name=None, rpath=None):
     if sys.platform == "darwin" and lib_name is not None:
         args += ["-Wl,-install_name,@rpath/%s.so" % lib_name]
```
So based on https://github.com/cocotb/cocotb/pull/1714#issuecomment-619625219 we can remove those also?

mtikekar replied:
We need to keep it. The default `install_name` is the path where the library got written by the compiler which will be some temporary build directory that pip uses.

mtikekar commented:
Reading through the build_libs file, I realized that a lot of the code is for building libraries that are not Python extensions. I was wondering if this will help: https://github.com/mdavidsaver/setuptools_dso

themperek commented:
Another possibility is #1387 but any solution needs some time to debug typically

eric-wieser commented:
@mtikekar, that looks very useful, thanks. I'd be keen to avoid maintaining our own code to build shared libraries. If you have time to attempt to put a patch together to use that, it would be appreciated!

mtikekar commented:
@themperek, I have pushed the smallest changes that get things working on my Mac setup. Hopefully, these can be merged without issues. 

In the first, I added libpython's path to rpath for libcocotb to find it. On Travis, you're using the Homebrew python whose libpython can be found in the hardcoded location set in its install_name (something like `/usr/local/opt/python/Frameworks/Python.framework/Versions/3.7/Python`). I am using conda python whose install name is `@rpath/libpython3.7.dylib`, so setting an rpath is necessary.

In the second commit, I added install_name to the vpi libraries so that verilator can work. Other simulators load the vpi library at runtime but verilator links it into the simulation executable. Without the right install_name, the executable looks for `build/lib.macosx-10.9-x86_64-3.7/cocotb/libs/libcocotbvpi_verilator.so` which it cannot find.

Finally, I cleaned up rpath at link-time so that `install_name_tool` is not needed.

themperek commented on cocotb_build_libs.py:
```diff
@@ -284,7 +272,7 @@ def _get_vpi_lib_ext(
             os.path.join(share_lib_dir, "vpi", "VpiCbHdl.cpp"),
         ],
-        extra_link_args=_extra_link_args(rpath="$ORIGIN"),
+        extra_link_args=_extra_link_args(lib_name=lib_name, rpath="$ORIGIN"),
```
Should we do the same for VHPI and FLI? Even if there is not simulator that supports this on Mac?

mtikekar replied:
Sure, I can make them all consistent.

elgorwi commented:
> For me it looks like a good change:
> 
>     * fixes issues on different osx setups
> 
>     * less hacky build_ext
> 
> 
> Let's wait for a few days. It would be good if @eric-wieser had a look too.
> 
> Should cleanup/squash commits.

Just tried this on my mac and installing now works without issues. 
 @eric-wieser @themperek Would be great to get this merged

mtikekar commented:
Do you need this squashed to a single commit?

On Wed, May 6, 2020, 2:32 AM elgorwi <notifications@github.com> wrote:

> For me it looks like a good change:
>
> * fixes issues on different osx setups
>
> * less hacky build_ext
>
> Let's wait for a few days. It would be good if @eric-wieser
> <https://github.com/eric-wieser> had a look too.
>
> Should cleanup/squash commits.
>
> Just tried this on my mac and installing now works without issues.
> @eric-wieser <https://github.com/eric-wieser> @themperek
> <https://github.com/themperek> Would be great to get this merged
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/cocotb/cocotb/pull/1714#issuecomment-624304317>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABCYIG7ZQUXC6T224WKRCGDRQB5GBANCNFSM4MRLCEQA>
> .
>

eric-wieser commented:
We can squash to a single commit through the github UI. If the commits are reasonably well segregated, then we tend to rebase-merge.

eric-wieser commented on cocotb_build_libs.py:
```diff
@@ -47,12 +47,10 @@ def build_extension(self, ext):
                 "-add_rpath",
                 "@loader_path/libs",
+                "-add_rpath",
+                sysconfig.get_config_var("LIBDIR"),
```
I assume you combined these into a single invocation because it was tidier, not because it was necessary?

Would be great if this commit message could mention that.

eric-wieser replied:
Nevermind, I see this gets removed in a later commit anyway - guess we should squash this PR when we merge.

eric-wieser commented on cocotb_build_libs.py:
```diff
@@ -228,14 +212,17 @@ def _get_common_lib_ext(include_dir, share_lib_dir):
     #
     #  libcocotb
     #
+    link_args = _extra_link_args(lib_name="libcocotb", rpath="$ORIGIN")
+    for lib_dir in python_lib_dirs:
+        link_args += _extra_link_args(rpath=lib_dir)
```
Perhaps instead the `rpath` argument to `_extra_link_args` should be a list, making this:

```suggestion
    link_args = _extra_link_args(lib_name="libcocotb", rpath=["$ORIGIN"] + python_lib_dirs)
    for lib_dir in python_lib_dirs:
        link_args += _extra_link_args(rpath=lib_dir)
```

mtikekar replied:
Good idea. Changed `rpath` to `rpaths` list.

eric-wieser commented on cocotb_build_libs.py:
```diff
@@ -386,7 +375,7 @@ def get_ext():
                     os.path.join(share_lib_dir, "fli", "FliObjHdl.cpp"),
                 ],
-                extra_link_args=_extra_link_args(rpath="$ORIGIN"),
+                extra_link_args=_extra_link_args(lib_name="libcocotbfli_modelsim", rpath="$ORIGIN"),
```
Can we create a `lib_name` variable for consistency here too?

mtikekar replied:
Done

mtikekar commented:
Did the suggested cleanup.

I think some more changes are needed to make a wheel out of this to put on pypi. The problem is with how `simulator` and `libcocotb` find `libpython`. Right now, on Linux and Mac, they're given a `rpath` to the location of `libpython` that was used in the build environment.

For `simulator`, I was looking at how `numpy` does it. And their extension libraries don't even link to `libpython` (on both Linux and Mac). On Mac, I think it works because of an `-undefined dynamic_lookup` link option which allows symbols from `libpython` to be found at load time. And at load time, the `python` interpreter has already loaded `libpython` before it loads the extensions, so the extensions get their symbols that way. On Linux, something similar might be at work. 

This is just my guess, by the way. It will need some more investigation. I could be completely wrong. In any case, the above mechanism will not work for `libcocotb` as that is not a python extension but a library that embeds python.

themperek commented on cocotb_build_libs.py:
```diff
@@ -137,7 +119,7 @@ def _gen_import_libs(self, def_dir):
 
 
-def _extra_link_args(lib_name=None, rpath=None):
+def _extra_link_args(lib_name=None, rpaths=[]):
```
```suggestion
def _extra_link_args(lib_name=None, rpaths=None):
```
I guess we should avoid `[]` here see: https://docs.python-guide.org/writing/gotchas/#mutable-default-arguments

themperek replied:
You are right doe not really apply here.

mtikekar commented:
I'm aware of that - it only matters if you mutate the value in the function
which we are not doing. But, to be safe, we can make it an empty tuple.

>
>
>

themperek commented on cocotb_build_libs.py:
```diff
@@ -235,7 +219,7 @@ def _get_common_lib_ext(include_dir, share_lib_dir):
         library_dirs=python_lib_dirs,
         sources=[os.path.join(share_lib_dir, "embed", "gpi_embed.cpp")],
-        extra_link_args=_extra_link_args(lib_name="libcocotb", rpath="$ORIGIN"),
+        extra_link_args=_extra_link_args(lib_name="libcocotb", rpaths=["$ORIGIN"] + python_lib_dirs),
```
So we have to add `python_lib_dirs` here because you get an error 
```
dyld: Library not loaded: @rpath/libpython3.7m.dylib
  Referenced from: .../lib/python3.7/site-packages/cocotb/libs/libcocotb.so
  Reason: image not found
```
Why does it work when using `brew`?

mtikekar replied:
brew python is always installed in a fixed global location. So the `install_name` of `libpython` contains the full path and any library that links to `libpython` will look for it in the full path. So, rpath is not needed. Compare the output of `otool -L libcocotb.so` (Mac-equivalent of `ldd`) on the master branch for brew vs conda pythons:

For brew python:
```
libcocotb.so:
        @rpath/libcocotb.so (compatibility version 0.0.0, current version 0.0.0)
        /usr/local/opt/python/Frameworks/Python.framework/Versions/3.7/Python (compatibility version 3.7.0, current version 3.7.0)
        @rpath/libgpilog.so (compatibility version 0.0.0, current version 0.0.0)
        @rpath/libcocotbutils.so (compatibility version 0.0.0, current version 0.0.0)
        /usr/lib/libc++.1.dylib (compatibility version 1.0.0, current version 400.9.0)
        /usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1252.50.4)
```

For conda python:
```
libcocotb.so:
        @rpath/libcocotb.so (compatibility version 0.0.0, current version 0.0.0)
        @rpath/libpython3.7m.dylib (compatibility version 3.7.0, current version 3.7.0)
        @rpath/libgpilog.so (compatibility version 0.0.0, current version 0.0.0)
        @rpath/libcocotbutils.so (compatibility version 0.0.0, current version 0.0.0)
        @rpath/libc++.1.dylib (compatibility version 1.0.0, current version 1.0.0)
        /usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1252.50.4)
```

Hence, `rpath` needs `python_lib_dirs` for conda python.

themperek replied:
Thank you for your reply.
It seems adds up complication if we ever would want to distribute binary package (as you commented before) but this is one of many issues ...

themperek commented:
Thank you @mtikekar for your contribution.
